extensions/amp-story-shopping/0.1/amp-story-shopping-config.js

```diff
@@ -1,19 +1,199 @@
+  return (field, value) => {
+    if (!validateConfig(value, validation, field)) {
+      throw Error(
+        `Value for field '${field}' is not vaild, see the error messages above for details`
+      );
+    }
+  };
+}
+
+/**
+ * Validates an Array of Objects using the validateConfig function.
+ * @param {?Object=} validation
+ * @return {boolean}
+ */
+function createValidateConfigArray(validation) {
+  return (field, value) => {
+    let isValid = true;
+    for (const item of value) {
+      isValid &&= validateConfig(item, validation, field);
+    }
+    if (!isValid) {
+      `Value for field '${field}' is not vaild, see the error messages above for details`;
+    }
+  };
+}
+
+/**
+ * Validates if a required field exists for shopping config attributes
+ * @param {string} field
+ * @param {?string=} value
+ */
+export function validateRequired(field, value) {
+  if (value === undefined) {
+    throw Error(`Field ${field} is required.`);
+  }
+}
+
+/**
+ * Validates if string type for shopping config attributes
+ * @param {string} field
+ * @param {?string=} str
+ */
+export function validateString(field, str) {
+  if (typeof str !== 'string') {
+    throw Error(`${field} ${str} is not a string`);
+  }
+}
+
+/**
+ * Validates if it is a valid HTML ID for shopping config attributes
+ * @param {string} field
+ * @param {?string=} id
+ */
+export function validateHTMLId(field, id) {
+  const checkID = /^[A-Za-z]+[\w\-\:\.]*$/;
+  if (!checkID.test(id)) {
+    throw Error(`${field} ${id} is not a valid HTML Id`);
+  }
+}
+
+/**
+ * Validates number in shopping config attributes
+ * @param {string} field
+ * @param {?number=} number
+ */
+export function validateNumber(field, number) {
+  if (
+    (typeof number === 'string' && !/^[0-9.,]+$/.test(number)) ||
+    (typeof number !== 'string' && typeof number !== 'number')
+  ) {
+    throw Error(`Value ${number} for field ${field} is not a number`);
+  }
+}
+
+/**
+ * Validates currency in shopping config attributes
+ * @param {string} field
+ * @param {?string=} curr
+ */
+export function validateCurrency(field, curr) {
+  // This will throw an error on invalid currency codes.
+  Intl.NumberFormat('en-EN', {currency: curr}).format(0);
+}
+
+/**
+ * Validates url of shopping config attributes
+ * @param {string} field
+ * @param {?Array<string>=} url
+ */
+export function validateURLs(field, url) {
+  if (url === undefined) {
+    return;
+  }
+
+  const urls = Array.isArray(url) ? url : [url];
+
+  urls.forEach((url) => {
+    assertHttpsUrl(url, `amp-story-shopping-config ${field}`);
+  });
+}
+
+/**
+ * Validates the shopping config of a single product.
+ * @param {!ShoppingConfigDataDef} shoppingConfig
+ * @param {!Object<string, !Array<function>>} validationObject
+ * @param {?string} optParentFieldName
+ * @return {boolean}
+ */
+export function validateConfig(
+  shoppingConfig,
+  validationObject = productValidationConfig,
+  optParentFieldName
```

Comment:
For optional values, it's preferred to remove the "opt" prefix and instead give it a default value of undefined. 

Also, could you add a description for this field? It's may not be immediately obvious what it represents

Reply:
[In Javascript, default values are already initialized to undefined](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Default_parameters#description)

I'll add a description.